vllm/model_executor/layers/mamba/ops/mamba_ssm.py

```diff
@@ -99,22 +110,52 @@ def _selective_scan_update_kernel(
     HAS_D: tl.constexpr,
     HAS_Z: tl.constexpr,
     HAS_STATE_BATCH_INDICES: tl.constexpr,
+    INPLACE_FINAL_STATE: tl.constexpr,
+    IS_SPEC_DECODING: tl.constexpr,
+    IS_VARLEN: tl.constexpr,
     BLOCK_SIZE_DSTATE: tl.constexpr,
 ):
     pid_m = tl.program_id(axis=0)
     pid_b = tl.program_id(axis=1)
     pid_h = tl.program_id(axis=2)
 
+    if IS_VARLEN:
+        bos = tl.load(cu_seqlens_ptr + pid_b).to(tl.int64)
+        eos = tl.load(cu_seqlens_ptr + pid_b + 1).to(tl.int64)
+        seq_len = eos - bos
+
+        if seq_len == 0:
+            return
+    else:
+        bos = pid_b * T
+        eos = bos + T
+        seq_len = T
```

Comment:
Could you explain this part a bit? I'm a bit confused about what's going on.

I thought that `cu_seqlens_ptr` would always be set now in vLLM. And what is T? 

My guess is that the speculative decoding changes to this kernel would only be used for pure decode cases, and T is the number of speculated tokens, but I'm not sure

Reply:
Appreciate the explanation. That helps explain what's going on quite a bit.

Should we ditch the variable `T` in order to simplify the code?